pkg/core/bundle_list.go

```diff
@@ -12,52 +13,128 @@ import (
 )
 
 const (
-	maxBundlesToList = 1000000
+	maxBundlesToList  = 1000000
+	typicalBundlesNum = 1000
 )
 
+type bundleChan struct {
+	bundle model.BundleDescriptor
+	err    error
+}
+
+type bundlesChan struct {
+	bundles []model.BundleDescriptor
+	err     error
+}
+
+// ListBundles returns the list of bundle descriptors from a repo.
+//
 // TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store) ([]model.BundleDescriptor, error) {
+func ListBundles(repo string, store storage.Store, opts ...BundleOption) ([]model.BundleDescriptor, error) {
+	var concurrencyFactor int
+	b := defaultBundle()
+	for _, bApply := range opts {
+		bApply(&b)
+	}
+	concurrencyFactor = b.concurrentBundleList
+
 	// Get a list
 	e := RepoExists(repo, store)
 	if e != nil {
 		return nil, e
 	}
+
+	// this call is not made async yet
 	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxBundlesToList)
 	if err != nil {
 		return nil, err
 	}
-	bds := make([]model.BundleDescriptor, 0)
-	for _, k := range ks {
+
+	var wg sync.WaitGroup
+
+	result := make(chan bundlesChan)
+	bundle := make(chan bundleChan, concurrencyFactor)
+
+	wg.Add(1)
+	go func(bundle <-chan bundleChan, result chan<- bundlesChan, wg *sync.WaitGroup) {
+		// coalesce result
+		defer wg.Done()
+		defer close(result)
+		bds := make([]model.BundleDescriptor, 0, typicalBundlesNum) // preallocate some typical number of bundles, e.g. 1000
+		for observed := range bundle {
+			if observed.err != nil {
+				result <- bundlesChan{err: observed.err}
+				return
+			}
+			bds = append(bds, observed.bundle)
+		}
+		result <- bundlesChan{bundles: bds}
+	}(bundle, result, &wg)
+
+	wg.Add(1)
+	go func(bundle chan<- bundleChan, wg *sync.WaitGroup) {
+		// start a bounded number of workers fetching descriptors.
+		// Limitation: this currently spans the whole set of keys, even though an early error has been met
+		defer wg.Done()
+		var workers sync.WaitGroup
+		keys := make(chan string, concurrencyFactor)
+
+		for count, k := range ks {
+			if count < concurrencyFactor {
+				workers.Add(1)
```

Comment:
how about spinning up the `getBundleAsync` worker goroutines synchronously before starting to send tasks (`keys` channel values) out via this goroutine?  supposing a `MinInt(int, int) int` util, this could look like
```
keys := make(chan string)
var workers sync.WaitGroup
numWorkers := MinInt(concurrencyFactor, len(ks))
for i := 0; i < numWorkers; i++ {
  workers.Add(1)
  go getBundleAsync(repo, store, keys, bundle, &workers)
}
```
then, with the additional change to collect results synchronously, the buffers can be removed from both the `keys` and `bundle` channels.  the current opinion is that channels should be unbuffered if at all possible for now in order to favor correctness.  we can add buffering for performance later on as necessary.

Reply:
Yes I am considering that.

Reply:
ok. So:
- [x] Going unbuffered
- [x] Spinning workers beforehand
- [x] Removing extra data copy to channel
- [ ] "synchronous" push/fetch: well, I am sticking to "watcher" single goroutines to prevent blocking sends